### PR TITLE
Partial implementation of retrieveBlobs in the public API

### PR DESCRIPTION
**Pull request by dcoutts**
It is partial because there is no support for BlobRefs in the write buffer yet.

As such it doesn't make a lot of sense to test in this interim state, so this PR does not enable tests for the public API. This can follow once we incorporate WriteBuffer support.

### CODE REVIEW

File: src/Database/LSMTree/Common.hs [hunk 1]
```diff
@@ -243,4 +243,4 @@ listSnapshots (Session sesh) = Internal.listSnapshots sesh
 -- TODO: get rid of the @m@ parameter?
 type BlobRef :: (Type -> Type) -> Type -> Type
 type role BlobRef nominal nominal
-data BlobRef m blob = forall h. Typeable h => BlobRef (Internal.BlobRef (Internal.Run m h))
+data BlobRef m blob = forall h. Typeable h => BlobRef (Internal.BlobRef (Internal.Run m (Handle h)))
```
Comment (phadej) [inline src/Database/LSMTree/Common.hs:246]: Shouldn't this be

```haskell
BlobRef (Internal.BlobRef (Handle h)`
```

or is Run -> Handle happening later?

File: src/Database/LSMTree/Internal.hs [hunk 2]
```diff
@@ -680,6 +684,65 @@ updates resolve es th = do
...
+            bufOffs = V.scanl (+) 0 (V.map blobRefSpanSize refs)
+        buf <- P.newPinnedByteArray bufSize
+        let ioops = V.zipWith (Run.readBlobIOOp buf) bufOffs refs
+            hbio  = sessionHasBlockIO seshEnv
+
+        -- Submit the IOOps all in one go:
+        _ <- FS.submitIO hbio ioops
+        -- We do not need to inspect the results because IO errors are
+        -- thrown as exceptions, and the result is just the read length
+        -- which is already known. Short reads can't happen here.
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs:717]: We should probably change [this code](https://github.com/well-typed/blockio-uring/blob/main/src/System/IO/BlockIO/URing.hs#L111-L140). Do the pattern synonyms do anything? If we can always expect `IOResult` to contain a non-negative integer?

We are currently explicitly handling the case where an `IOResult` can contain an error number in https://github.com/IntersectMBO/lsm-tree/blob/main/blockio-api/src-linux/System/FS/BlockIO/Async.hs#L75-L99. It should probably be a catch+rethrow instead

File: src/Database/LSMTree/Common.hs (hunk 1 above)
Comment (dcoutts) [reply src/Database/LSMTree/Common.hs:242]: Yes, not in this PR. In this PR I'm just following the existing strategy.

I plan to make a separate PR that changes the BlobRef representation to use a pair of a handle and a ref count, and hopefully that can be our unified representation for blob refs into both runs and the write buffer.

File: src-control/Control/RefCount.hs [hunk 3]
```diff
@@ -65,19 +66,58 @@ mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
...
+-- This amounts to trying to increase the reference count, but if it is already
+-- zero then this will fail. And unlike with 'addReference' where such failure
+-- would be a programer error, this corresponds to the case when the thing the
+-- reference count is tracking has been closed already.
+--
+-- The result is @True@ when a strong reference has been obtained and @False@
+-- when upgrading fails.
+--
+upgradeWeakReference :: PrimMonad m => RefCounter m -> m Bool
+upgradeWeakReference RefCounter{countVar} = do
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src-control/Control/RefCount.hs:103]: Yes indeed.

File: src/Database/LSMTree/Internal.hs [hunk 4]
```diff
@@ -680,6 +684,65 @@ updates resolve es th = do
...
+      V.imapM_ (\i ref -> do
+                  ok <- Run.upgradeWeakReference (blobRefRun ref)
+                  when (not ok) $ do
+                    -- drop refs on the previous ones taken successfully so far
+                    V.mapM_ (Run.removeReference . blobRefRun) (V.take i refs)
+                    throwIO ErrBlobRefInvalid
+               ) refs
+
+    removeReferences =
+      V.mapM_ (Run.removeReference . blobRefRun) refs
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:744]: I think the existing approach and API is fine. It's really a programmer error if the blob ref becomes invalidated. So it doesn't need to have a graceful fallback, especially when that would complicate the API in the normal case. Throwing an exception is fine IMHO.

File: src/Database/LSMTree/Internal.hs [hunk 5]
```diff
@@ -680,6 +684,65 @@ updates resolve es th = do
...
+    -- open using a reference count. So we have to upgrade our weak
+    -- reference to a (normal) strong reference while we read the blob, to
+    -- ensure it is not closed under our feet.
+    upgradeWeakReferences =
+      V.imapM_ (\i ref -> do
+                  ok <- Run.upgradeWeakReference (blobRefRun ref)
+                  when (not ok) $ do
+                    -- drop refs on the previous ones taken successfully so far
+                    V.mapM_ (Run.removeReference . blobRefRun) (V.take i refs)
+                    throwIO ErrBlobRefInvalid
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:740]: Yes fair enough.

File: src/Database/LSMTree/Internal.hs (hunk 2 above)
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:728]: The `blockio-uring` lib does not throw exceptions. The `IOError` and `IOResult` patters do what you'd expect.

It's our wrapper in `blockio-api` that changes the `IOResult` to have a single constructor for the successful case. And so the instance for `blockio-uring` does that by throwing exceptions.

The code above is using our wrapper, and the comment refers to that, not to the underlying `blockio-uring`.

File: src/Database/LSMTree/Internal.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs]: Done.

File: src/Database/LSMTree/Internal.hs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE CPP       #-}
 {-# LANGUAGE DataKinds #-}
+{- HLINT ignore "Use unless" -}
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs:3]: You can add this ignore `hlint.yaml`

**Review by jorisdral — APPROVED** [review]

**Pull request merged**